### DISCUSSION TITLE
Plans to expand networks?

### DISCUSSION THREAD

**Discussion by plainerman** [rust-bitcoin/rust-bitcoin#885] [Q&A] [answered]
The library is very written very versatile, and many parts could be used for different coins, especially bitcoin forks (e.g. litecoin).

The current structure makes it difficult to add new networks, which should be sufficient to support many parts of add additional coins (e.g. address generation etc.).

Are there any plans to support derivative coins? Or possibly a way to extend the `Network` enum without forking the project?

What are your opinions on this topic?

**Comment by apoelstra** [comment, marked as answer]
This library was started in the "clonecoin" era when most altcoins were just forks of Bitcoin Core (then just "bitcoin") with all the strings changed, a random and insufficient set of numeric constants changed, and no ongoing maintenance. In that context we were willing to add new sets of constants to our `Network` enum and other such places, but unwilling to meaningfully change the architecture of the library.

Even this small level of support has caused us problems, primarily due to the lack of `#[non_exhaustive]` markers on our enums, which have meant that any new constants wourld be breaking changes and our downstream users' code would be potentially complicated by their addition. To isolate this, there have been a few attempts to create a "bitcoin-constants" crate which would isolate the churn, but IMHO we were never able to find a clean enough architecture to justify the review burden and bikeshedding.

Today the mainstream altcoins all have a meaningfully different architecture from Bitcoin Core, which makes this sort of thing less useful. (Without digressing too much to rant about litecoin, I would not consider it "mainstream".)

Anyway we should revisit this after the upcoming MSRV bump. It may be that with the `#[non_exhaustive]` attribute, more modern and consistent Rust idioms, and @tcharding working full-time on this, we can come up with a bitcoin-constants crate that we'll all be happy with, and where we can stick extra constants in for random clonecoins without being too disruptive.

**Reply by Kixunil** [reply]
We've turned `max_money` function into a constant. Which implies no coins that change money supply. I guess only scams like bcrash remain? I'm not interested in supporting those.

If someone wants a completely independent `bitcoin-and-shitcoins` crate that's fine but I'm against adding anything, even constants into `bitcoin`. If compatibility with `bitcoin` is desired one may reverse the dependencies:

```rust
#[non_exhaustive]
pub enum Network {
    Bitcoin(bitcoin::Network),
    Shitcoin0,
    Shitcoin1,
    // ...
    ShitcoinN,
}
```

To avoid anyone wasting their time, I promise to NACK any shitcoin-adding PR without review unless one of the current maintainers has some **very** good argument. (This is not to support argument from authority but to not waste time reading a likely-broken argument from random strangers.)

**Comment by TheBlueMatt** [comment]
Litecoin is now no longer a full-on clone of Bitcoin, so even for it we can't just add a constant and call it done. I'm not sure if there are *any* clonecoins left that have more than about five users and are really just pure clonecoins.

**Comment by tcharding** [comment]
Collaboration between projects is good but, in my opinion, to be fair to those giving their time to working on this project it would have to be done in a way that does not introduce too much maintenance burden. Additionally, flyby contributions tend to burden the maintainers of a project, so it would be nice if we had some evidence that adding clonecoin support was not just a flyby contribution that the rust-bitcoin devs would then be left to maintain. That said, `rust-bitcoin` is pure open source, you should definitely fork it, use it, hack on it, and try to push your changes back upstream to reduce _your_ maintenance burden and improve the project for everyone :)

**Comment by dr-orlovsky** [comment]
As a maintainer I can clearly state that I do not have a bit of time to spent on blockchain-level shitcoins or increasing compatibility with them.
